Framework/Geometry/src/Instrument/ParameterMap.cpp

```diff
@@ -1199,20 +1227,30 @@ size_t ParameterMap::componentIndex(const ComponentID componentId) const {
   return m_componentInfo->indexOf(componentId);
 }
 
-/// Only for use by ExperimentInfo. Sets the pointer to the DetectorInfo.
-void ParameterMap::setDetectorInfo(
-    boost::shared_ptr<const Beamline::DetectorInfo> detectorInfo) {
-  m_detectorInfo = std::move(detectorInfo);
-}
-
-void ParameterMap::setComponentInfo(
-    boost::shared_ptr<const Geometry::ComponentInfo> componentInfo) {
-  m_componentInfo = std::move(componentInfo);
-}
-
 /// Only for use by Instrument. Sets the pointer to the owning instrument.
 void ParameterMap::setInstrument(const Instrument *instrument) {
+  if (instrument == m_instrument)
+    return;
+  if (!instrument) {
+    m_componentInfo = nullptr;
+    m_detectorInfo = nullptr;
+    return;
+  }
+  if (m_instrument)
+    throw std::logic_error("ParameterMap::setInstrument: Cannot change "
+                           "instrument once it has been set.");
+  if (instrument->isParametrized())
+    throw std::logic_error("ParameterMap::setInstrument must be called with "
+                           "base instrument, not a parametrized instrument");
   m_instrument = instrument;
+
+  const auto parInstrument = ParComponentFactory::createInstrument(
+      boost::shared_ptr<const Instrument>(m_instrument, NoDeleting()),
+      boost::shared_ptr<ParameterMap>(this, NoDeleting()));
+
+  InstrumentVisitor visitor(parInstrument);
```

Comment:
Overall, neat. I was thinking we could shield this even further with a non-member on the `InstrumentVisitor`
Something that would compress the following down to.
```cpp
std::tie(m_componentInfo, m_detectorInfo) = makeWrappers(parInstrument);
```